From 000e5368b1cec21ea8114fcfd763b7fbb0c179b4 Mon Sep 17 00:00:00 2001
From: "hanliang.xiong" <hanliang.xiong@amlogic.com>
Date: Fri, 24 Mar 2023 09:36:04 +0800
Subject: [PATCH] Fix hwclock -s fail on glibc new version.

Signed-off-by: hanliang.xiong <hanliang.xiong@amlogic.com>
---
 util-linux/hwclock.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/util-linux/hwclock.c b/util-linux/hwclock.c
index dc97d8fb4..8df14748c 100644
--- a/util-linux/hwclock.c
+++ b/util-linux/hwclock.c
@@ -131,7 +131,7 @@ static void to_sys_clock(const char **pp_rtcname, int utc)
 
 	tv.tv_sec = read_rtc(pp_rtcname, NULL, utc);
 	tv.tv_usec = 0;
-	if (settimeofday(&tv, &tz))
+	if (settimeofday(&tv, NULL))
 		bb_simple_perror_msg_and_die("settimeofday");
 }
 
@@ -283,7 +283,7 @@ static void set_system_clock_timezone(int utc)
 	gettimeofday(&tv, NULL);
 	if (!utc)
 		tv.tv_sec += tz.tz_minuteswest * 60;
-	if (settimeofday(&tv, &tz))
+	if (settimeofday(&tv, NULL))
 		bb_simple_perror_msg_and_die("settimeofday");
 }
 
-- 
2.25.1

