From c0723d436eedc045f8444dad11efb201a92092bc Mon Sep 17 00:00:00 2001
From: Dezhi Kong <dezhi.kong@amlogic.com>
Date: Mon, 16 Nov 2020 10:40:03 +0800
Subject: [PATCH] weston: add gbm modifier support [1/1]

---
 configure.ac | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/configure.ac b/configure.ac
index 0cc559d..fddf383 100644
--- a/configure.ac
+++ b/configure.ac
@@ -248,6 +248,11 @@ if test x$enable_drm_compositor = xyes; then
   PKG_CHECK_MODULES(DRM_COMPOSITOR_GBM, [gbm],
 		    [AC_DEFINE([HAVE_GBM_FD_IMPORT], 1, [gbm supports import with modifiers])],
 		    [AC_MSG_WARN([GBM does not support dmabuf import, will omit that capability])])
+  AC_ARG_ENABLE(gbm-modifiers, [  --enable-gbm-modifiers],enable_gbm_modifiers=yes,
+		enable_gbm_modifiers=no)
+  if test x$enable_gbm_modifiers = xyes; then
+	AC_DEFINE([HAVE_GBM_MODIFIERS], 1, [GBM supports modifiers])
+  fi
 fi
 
 AC_ARG_ENABLE(output-dynamic-switch, [  --enable-output-dynamic-switch],,
-- 
2.28.0

