diff --git a/binc/agent.c b/binc/agent.c
index ee7d35a..fbb953c 100644
--- a/binc/agent.c
+++ b/binc/agent.c
@@ -40,6 +40,7 @@ struct binc_agent {
     guint registration_id;
     AgentRequestAuthorizationCallback request_authorization_callback;
     AgentRequestPasskeyCallback request_passkey_callback;
+    AgentPasskeyNotifyCallback passkey_notify_callback;
 };
 
 void binc_agent_free(Agent *agent) {
@@ -114,6 +115,10 @@ static void bluez_agent_method_call(GDBusConnection *conn,
         g_dbus_method_invocation_return_value(invocation, NULL);
     } else if (g_str_equal(method, "RequestConfirmation")) {
         g_variant_get(params, "(ou)", &object_path, &pass);
+        Device *device = binc_adapter_get_device_by_path(adapter, object_path);
+        if (agent->passkey_notify_callback != NULL) {
+            agent->passkey_notify_callback(device, pass);
+        }
         g_free(object_path);
         log_debug(TAG, "request confirmation for %u", pass);
         g_dbus_method_invocation_return_value(invocation, NULL);
@@ -299,6 +304,12 @@ void binc_agent_set_request_passkey_cb(Agent *agent, AgentRequestPasskeyCallback
     agent->request_passkey_callback = callback;
 }
 
+void binc_agent_set_passkey_notify_cb(Agent *agent, AgentPasskeyNotifyCallback callback) {
+    g_assert(agent != NULL);
+    g_assert(callback != NULL);
+    agent->passkey_notify_callback = callback;
+}
+
 const char *binc_agent_get_path(const Agent *agent) {
     g_assert(agent != NULL);
     return agent->path;
diff --git a/binc/agent.h b/binc/agent.h
index b6d37ed..06bb52f 100644
--- a/binc/agent.h
+++ b/binc/agent.h
@@ -34,6 +34,7 @@ extern "C" {
 typedef gboolean (*AgentRequestAuthorizationCallback)(Device *device);
 
 typedef guint32 (*AgentRequestPasskeyCallback)(Device *device);
+typedef void (*AgentPasskeyNotifyCallback)(Device *device,guint32 passkey);
 
 typedef enum {
     DISPLAY_ONLY, DISPLAY_YES_NO, KEYBOARD_ONLY, NO_INPUT_NO_OUTPUT, KEYBOARD_DISPLAY
@@ -45,6 +46,8 @@ void binc_agent_set_request_authorization_cb(Agent *agent, AgentRequestAuthoriza
 
 void binc_agent_set_request_passkey_cb(Agent *agent, AgentRequestPasskeyCallback callback);
 
+void binc_agent_set_passkey_notify_cb(Agent *agent, AgentPasskeyNotifyCallback callback);
+
 void binc_agent_free(Agent *agent);
 
 const char *binc_agent_get_path(const Agent *agent);
