From 3f39197f30b53792b44f3681199ada1003ddeff2 Mon Sep 17 00:00:00 2001
From: "tianwei.ge" <tianwei.ge@thirdreality.ai>
Date: Tue, 6 Jan 2026 11:08:55 +0800
Subject: [PATCH] Fix quantization to use int8 instead of uint8

---
 pymicro_wakeword/microwakeword.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pymicro_wakeword/microwakeword.py b/pymicro_wakeword/microwakeword.py
index d1fc411..2dff7ee 100644
--- a/pymicro_wakeword/microwakeword.py
+++ b/pymicro_wakeword/microwakeword.py
@@ -193,7 +193,7 @@ class MicroWakeWord(TfLiteWakeWord):
         quant_features = np.round(
             np.concatenate(self._features, axis=1) / self.input_scale
             + self.input_zero_point
-        ).astype(np.uint8)
+        ).astype(np.int8)
 
         # Stride instead of rolling
         self._features.clear()
-- 
2.40.1

